sky/serve/core.py

```diff
@@ -262,6 +263,116 @@ def up(
+        controller_type=controller_utils.Controllers.SKY_SERVE_CONTROLLER,
+        stopped_message=
+        'Service controller is stopped. There is no service to update. '
+        f'To spin up a new service, use {backend_utils.BOLD}'
+        f'sky serve up{backend_utils.RESET_BOLD}',
+        non_existent_message='Service does not exist. '
+        'To spin up a new service, '
+        f'use {backend_utils.BOLD}sky serve up{backend_utils.RESET_BOLD}',
+    )
+
+    if handle is None or handle.head_ip is None:
+        # The error message is already printed in
+        # backend_utils.is_controller_up
+        # TODO(zhwu): Move the error message into the exception.
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ClusterNotUpError(message='',
+                                               cluster_status=cluster_status)
+
+    backend = backend_utils.get_backend_from_handle(handle)
+    assert isinstance(backend, backends.CloudVmRayBackend)
+
+    code = serve_utils.ServeCodeGen.get_service_status([service_name])
+    returncode, serve_status_payload, stderr = backend.run_on_head(
+        handle,
+        code,
+        require_outputs=True,
+        stream_logs=False,
+        separate_stderr=True)
+    try:
+        subprocess_utils.handle_returncode(returncode,
+                                           code, 'Failed to get service status '
+                                           'when update service',
+                                           stderr,
+                                           stream_logs=True)
+    except exceptions.CommandError as e:
+        raise RuntimeError(e.error_msg) from e
+
+    service_statuses = serve_utils.load_service_status(serve_status_payload)
+    if len(service_statuses) == 0:
+        with ux_utils.print_exception_no_traceback():
+            raise RuntimeError(f'Cannot find service {service_name!r}.'
+                               f'To spin up a service, use {backend_utils.BOLD}'
+                               f'sky serve up{backend_utils.RESET_BOLD}')
+
+    if len(service_statuses) > 1:
+        with ux_utils.print_exception_no_traceback():
+            raise RuntimeError(
+                f'Multiple services found for {service_name!r}. ')
+    service_record = service_statuses[0]
+    prompt = None
+    if (service_record['status'] == serve_state.ServiceStatus.CONTROLLER_FAILED
+       ):
+        prompt = (f'Service {service_name!r} has a failed controller. '
+                  'Please clean up the service and try again.')
+    elif (service_record['status'] == serve_state.ServiceStatus.CONTROLLER_INIT
+         ):
+        prompt = (f'Service {service_name!r} is still initializing '
+                  'its controller. Please try again later.')
+    if prompt is not None:
+        with ux_utils.print_exception_no_traceback():
+            raise RuntimeError(prompt)
+
+    version = service_record['version']
```

Comment:
This can cause concurrency issue, e.g., if there are two updates called, they will have the same version number.
1. `sky serve update`
2. Before the first step updates the database on the controller, call `sky serve update` again.


We probably want to refer to the implementation of the `job_id`, where we first ask for the latest `version+1` on the controller by using the atomic increment of the version in the database, so that multiple updates will have different version number and our update on the controller should only execute the update with the version number that is `>=` the version number in the database.

https://github.com/skypilot-org/skypilot/blob/c1f28bcb630b60f9a22d2303119f7ce62e700de3/sky/backends/cloud_vm_ray_backend.py#L3332-L3333

https://github.com/skypilot-org/skypilot/blob/c1f28bcb630b60f9a22d2303119f7ce62e700de3/sky/skylet/job_lib.py#L274-L288

https://github.com/skypilot-org/skypilot/blob/c1f28bcb630b60f9a22d2303119f7ce62e700de3/sky/skylet/job_lib.py#L60

Reply:
I see. Added